app/src/main/java/org/thoughtcrime/securesms/database/helpers/PreKeyMigrationHelper.java

```diff
@@ -0,0 +1,226 @@
+            contentValues.put(SignedPreKeyDatabase.TIMESTAMP, signedPreKey.getTimestamp());
+            database.insert(SignedPreKeyDatabase.TABLE_NAME, null, contentValues);
+            Log.i(TAG, "Migrated signed prekey: " + signedPreKey.getId());
+          } catch (IOException | InvalidMessageException e) {
+            Log.w(TAG, e);
+            clean = false;
+          }
+        }
+      }
+    }
+
+    File oneTimePreKeyIndex = new File(getPreKeyDirectory(context), PreKeyIndex.FILE_NAME);
+    File signedPreKeyIndex  = new File(getSignedPreKeyDirectory(context), SignedPreKeyIndex.FILE_NAME);
+
+    if (oneTimePreKeyIndex.exists()) {
+      try {
+        InputStreamReader reader = new InputStreamReader(new FileInputStream(oneTimePreKeyIndex));
+        PreKeyIndex        index = JsonUtils.fromJson(reader, PreKeyIndex.class);
+        reader.close();
+
+        Log.i(TAG, "Setting next prekey id: " + index.nextPreKeyId);
+        TextSecurePreferences.setNextPreKeyId(context, index.nextPreKeyId);
+      } catch (IOException e) {
+        Log.w(TAG, e);
+      }
+    }
+
+    if (signedPreKeyIndex.exists()) {
+      try {
+        InputStreamReader reader = new InputStreamReader(new FileInputStream(signedPreKeyIndex));
+        SignedPreKeyIndex index  = JsonUtils.fromJson(reader, SignedPreKeyIndex.class);
+        reader.close();
+
+        Log.i(TAG, "Setting next signed prekey id: " + index.nextSignedPreKeyId);
+        Log.i(TAG, "Setting active signed prekey id: " + index.activeSignedPreKeyId);
+        TextSecurePreferences.setNextSignedPreKeyId(context, index.nextSignedPreKeyId);
+        TextSecurePreferences.setActiveSignedPreKeyId(context, index.activeSignedPreKeyId);
+      } catch (IOException e) {
+        Log.w(TAG, e);
+      }
+    }
+
+    return clean;
+  }
+
+  static void cleanUpPreKeys(@NonNull Context context) {
+    File   preKeyDirectory = getPreKeyDirectory(context);
+    File[] preKeyFiles     = preKeyDirectory.listFiles();
+
+    if (preKeyFiles != null) {
+      for (File preKeyFile : preKeyFiles) {
+        Log.i(TAG, "Deleting: " + preKeyFile.getAbsolutePath());
+        preKeyFile.delete();
+      }
+
+      Log.i(TAG, "Deleting: " + preKeyDirectory.getAbsolutePath());
+      preKeyDirectory.delete();
+    }
+
+    File   signedPreKeyDirectory = getSignedPreKeyDirectory(context);
+    File[] signedPreKeyFiles     = signedPreKeyDirectory.listFiles();
+
+    if (signedPreKeyFiles != null) {
+      for (File signedPreKeyFile : signedPreKeyFiles) {
+        Log.i(TAG, "Deleting: " + signedPreKeyFile.getAbsolutePath());
+        signedPreKeyFile.delete();
+      }
+
+      Log.i(TAG, "Deleting: " + signedPreKeyDirectory.getAbsolutePath());
+      signedPreKeyDirectory.delete();
+    }
+  }
+
+  private static byte[] loadSerializedRecord(File recordFile)
+      throws IOException, InvalidMessageException
+  {
+    FileInputStream fin           = new FileInputStream(recordFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fin*. The resource is referenced at lines 153, 159. The resource is closed at line 165. There are other execution paths that do not contain closure statements, for example, when an if condition at line 161 evaluates to true. Ensure *fin* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).